examples/cert-validation.cpp

```diff
@@ -0,0 +1,112 @@
+#include <mococrw/ca.h>
+#include <mococrw/basic_constraints.h>
+#include <mococrw/key_usage.h>
+
+using namespace mococrw;
+
+#define CERTS "../../tests/unit/test-certs/"
+
+int main()
+{
+    auto rootCA = X509Certificate::fromPEMFile(CERTS"root3.pem");
+    auto intermediateCA = X509Certificate::fromPEMFile(CERTS"root3.int1.pem");
+    auto intermediateCA_1 = X509Certificate::fromPEMFile(CERTS"root3.int1.int11.pem");
+
+    auto rootCRL = CertificateRevocationList::fromPEMFile(CERTS"root3.crl.pem");
+    auto rootCRL_empty = CertificateRevocationList::fromPEMFile(CERTS"root3.crl_empty.pem");
+    auto intermediateCRL = CertificateRevocationList::fromPEMFile(CERTS"root3.int1.crl.pem");
+
+    auto cert = X509Certificate::fromPEMFile(CERTS"root3.int1.cert12.pem");
+
+    /*
+            Basic case - certificate revoked
+    */
+    X509Certificate::VerificationContext ctx_1;
+    ctx_1.addTrustedCertificate(rootCA)
+       .addIntermediateCertificate(intermediateCA)
+       .addCertificateRevocationList(rootCRL)
+       .addCertificateRevocationList(intermediateCRL)
+       .enforceCrlsForAllCAs()
+       .enforceSelfSignedRootCertificate();
+
+    try {
+        cert.verify(ctx_1);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Failed to verify certificate.\n";
+        std::cerr << e.what() << '\n';
+    }
+
+    /*
+            Basic case - certificate not revoked
+    */
+    X509Certificate::VerificationContext ctx_2;
+    ctx_2.addTrustedCertificate(rootCA)
+       .addIntermediateCertificate(intermediateCA)
+       .addCertificateRevocationList(rootCRL_empty)
+       .addCertificateRevocationList(intermediateCRL)
+       .enforceCrlsForAllCAs()
+       .enforceSelfSignedRootCertificate();
+
+    try {
+        cert.verify(ctx_2);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Failed to verify certificate.\n";
+        std::cerr << e.what() << '\n';
+    }
+
+    /*
+            Root CA missing CRL and enforceCrlsForAllCAs is set
+    */
+    X509Certificate::VerificationContext ctx_3;
+    ctx_3.addTrustedCertificate(rootCA)
+       .addIntermediateCertificate(intermediateCA)
+       .addCertificateRevocationList(intermediateCRL)
+       .enforceCrlsForAllCAs()
+       .enforceSelfSignedRootCertificate();
+
+    try {
+        cert.verify(ctx_3);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Failed to verify certificate.\n";
+        std::cerr << e.what() << '\n';
+    }
+
+    /*
+            Root CA is not self signed
+    */
+    X509Certificate::VerificationContext ctx_4;
+    ctx_4.addTrustedCertificate(intermediateCA)
+       .addIntermediateCertificate(intermediateCA_1)
+       .addCertificateRevocationList(intermediateCRL)
+       .enforceSelfSignedRootCertificate();
+
+    try {
+        cert.verify(ctx_4);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Failed to verify certificate.\n";
+        std::cerr << e.what() << '\n';
+    }
```

Comment:
You/we may want to illustrate the option to do certificate validation for a certain date and time.